pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/TableDebugResource.java

```diff
@@ -0,0 +1,267 @@
+   *
+   * @param pinotHelixResourceManager Helix Resource Manager for the cluster
+   * @param tableName Name of the table
+   * @param tableType Type of the table (offline|realtime)
+   * @return TableDebugInfo
+   */
+  private TableDebugInfo debugTable(PinotHelixResourceManager pinotHelixResourceManager, String tableName,
+      TableType tableType) {
+    String tableNameWithType = TableNameBuilder.forType(tableType).tableNameWithType(tableName);
+
+    // Debug information for segments of the table.
+    List<TableDebugInfo.SegmentDebugInfo> segmentDebugInfos =
+        debugSegments(pinotHelixResourceManager, tableNameWithType);
+
+    // Debug information from brokers of the table.
+    List<TableDebugInfo.BrokerDebugInfo> brokerDebugInfos = debugBrokers(tableNameWithType);
+
+    // Debug information from servers of the table.
+    List<TableDebugInfo.ServerDebugInfo> serverDebugInfos =
+        debugServers(pinotHelixResourceManager, tableName, tableType);
+
+    // Table size summary.
+    TableDebugInfo.TableSizeSummary tableSizeSummary = getTableSize(tableNameWithType);
+
+    // Number of segments in the table.
+    IdealState idealState = _pinotHelixResourceManager.getTableIdealState(tableNameWithType);
+    int numSegments = (idealState != null) ? idealState.getPartitionSet().size() : 0;
+
+    return new TableDebugInfo(tableNameWithType, tableSizeSummary, segmentDebugInfos, serverDebugInfos,
+        brokerDebugInfos, _pinotHelixResourceManager.getServerInstancesForTable(tableName, tableType).size(),
+        _pinotHelixResourceManager.getBrokerInstancesForTable(tableName, tableType).size(), numSegments);
+  }
+
+  private TableDebugInfo.TableSizeSummary getTableSize(String tableNameWithType) {
+    TableSizeReader tableSizeReader =
+        new TableSizeReader(_executor, _connectionManager, _controllerMetrics, _pinotHelixResourceManager);
+    TableSizeReader.TableSizeDetails tableSizeDetails;
+    try {
+      tableSizeDetails = tableSizeReader
+          .getTableSizeDetails(tableNameWithType, _controllerConf.getServerAdminRequestTimeoutSeconds() * 1000);
+    } catch (Throwable t) {
+      tableSizeDetails = null;
+    }
+
+    return (tableSizeDetails != null) ? new TableDebugInfo.TableSizeSummary(tableSizeDetails.reportedSizeInBytes,
+        tableSizeDetails.estimatedSizeInBytes) : new TableDebugInfo.TableSizeSummary(-1, -1);
+  }
+
+  /**
+   * Helper method to debug segments. Computes differences between ideal state and external view for each segment.
+   *
+   * @param pinotHelixResourceManager Helix Resource Manager
+   * @param tableNameWithType Name of table with type
+   * @return Debug information for segments
+   */
+  private List<TableDebugInfo.SegmentDebugInfo> debugSegments(PinotHelixResourceManager pinotHelixResourceManager,
+      String tableNameWithType) {
+    ExternalView externalView = pinotHelixResourceManager.getTableExternalView(tableNameWithType);
```

Comment:
How about the segments in EV but not in IS?

Reply:
In my experience, this is only due to delay in deleting segments. So chose not to add it for now, to avoid confusion.